app/lua/lrotable.c

```diff
@@ -9,77 +9,139 @@
 #include "lobject.h"
 #include "lapi.h"
 
-/* Local defines */
-#define LUAR_FINDFUNCTION     0
-#define LUAR_FINDVALUE        1
-
-/* Externally defined read-only table array */
-extern const luaR_table *lua_rotable;
-
-/* Find a global "read only table" in the constant lua_rotable array */
-void* luaR_findglobal(const char *name, unsigned len) {
-  unsigned i;
-
-  if (c_strlen(name) > LUA_MAX_ROTABLE_NAME)
-    return NULL;
-  for (i=0; lua_rotable[i].name; i ++)
-    if (*lua_rotable[i].name != '\0' && c_strlen(lua_rotable[i].name) == len && !c_strncmp(lua_rotable[i].name, name, len)) {
-      return (void*)(lua_rotable[i].pentries);
+#define ALIGNED_STRING (__attribute__((aligned(4))) char *)
+#define LA_LINES 16
+#define LA_SLOTS 4
+//#define COLLECT_STATS
+
+/*
+ * All keyed ROtable access passes through luaR_findentry().  ROTables
+ * are simply a list of <key><TValue value> pairs.  The existing algo
+ * did a linear scan of this vector of pairs looking for a match.
+ *
+ * A N×M lookaside cache has been added, with a simple hash on the key's
+ * TString addr and the ROTable addr to identify one of N lines.  Each
+ * line has M slots which are scanned. This is all done in RAM and is
+ * perhaps 20x faster than the corresponding random Flash accesses which
+ * will cause flash faults.
+ *
+ * If a match is found and the table addresses match, then this entry is
+ * probed first. In practice the hit-rate here is over 99% so the code
+ * rarely fails back to doing the linear scan in ROM.
+ *
+ * Note that this hash does a couple of prime multiples and a modulus 2^X
+ * with is all evaluated in H/W, and adequately randomizes the lookup.
+ */
+#define HASH(a,b) (519*((size_t)(a)>>4) + 17*((size_t)(b)>>4))
+
+static struct {
+  unsigned hash;
+  unsigned addr:24;
+  unsigned ndx:8;
+} cache[LA_LINES][LA_SLOTS];
```

Comment:
Eeep! That's half a K right there, isn't it? :(
Could this be an optional feature, so those really tight on RAM can opt for the performance hit instead?

Reply:
That was the ½K ☺️ I can make move it from static to dynamic, but we'll always need it - or at least the cache if we want fast ROTable access. I can half the line size with my next step on this journey. 

I'll also do the claw- back on DNS on this patch so we don't lose net RAM.

Reply:
No don't make it dynamic. This is something which if it's in use, it should be statically allocated to ensure it's there when needed. Plus it makes it easier to reason about other RAM usage.

If you can reclaim RAM from elsewhere so the net available stays the same, that would be Very Good :) And if half the size is enough for the common scenarios, that sounds good too!